packages/beacon-node/src/chain/archiver/archiveStates.ts

```diff
@@ -27,102 +13,27 @@ export interface StatesArchiverOpts {
-   * ```
-   */
   async maybeArchiveState(finalized: CheckpointWithHex): Promise<void> {
-    const lastStoredSlot = await this.db.stateArchive.lastKey();
-    const lastStoredEpoch = computeEpochAtSlot(lastStoredSlot ?? 0);
-    const {archiveStateEpochFrequency} = this.opts;
-
-    if (finalized.epoch - lastStoredEpoch >= Math.min(PERSIST_TEMP_STATE_EVERY_EPOCHS, archiveStateEpochFrequency)) {
-      await this.archiveState(finalized);
-
-      // Only check the current and previous intervals
-      const minEpoch = Math.max(
-        0,
-        (Math.floor(finalized.epoch / archiveStateEpochFrequency) - 1) * archiveStateEpochFrequency
-      );
-
-      const storedStateSlots = await this.db.stateArchive.keys({
-        lt: computeStartSlotAtEpoch(finalized.epoch),
-        gte: computeStartSlotAtEpoch(minEpoch),
-      });
-
-      const statesSlotsToDelete = computeStateSlotsToDelete(storedStateSlots, archiveStateEpochFrequency);
-      if (statesSlotsToDelete.length > 0) {
-        await this.db.stateArchive.batchDelete(statesSlotsToDelete);
-      }
-
-      // More logs to investigate the rss spike issue https://github.com/ChainSafe/lodestar/issues/5591
-      this.logger.verbose("Archived state completed", {
-        finalizedEpoch: finalized.epoch,
-        minEpoch,
-        storedStateSlots: storedStateSlots.join(","),
-        statesSlotsToDelete: statesSlotsToDelete.join(","),
-      });
-    }
+    await this.archiveState(finalized);
   }
 
   /**
    * Archives finalized states from active bucket to archive bucket.
    * Only the new finalized state is stored to disk
    */
   async archiveState(finalized: CheckpointWithHex): Promise<void> {
-    // starting from Mar 2024, the finalized state could be from disk or in memory
-    const finalizedStateOrBytes = await this.regen.getCheckpointStateOrBytes(finalized);
-    const {rootHex} = finalized;
-    if (!finalizedStateOrBytes) {
-      throw Error(`No state in cache for finalized checkpoint state epoch #${finalized.epoch} root ${rootHex}`);
-    }
-    if (finalizedStateOrBytes instanceof Uint8Array) {
-      const slot = getStateSlotFromBytes(finalizedStateOrBytes);
-      await this.db.stateArchive.putBinary(slot, finalizedStateOrBytes);
-      this.logger.verbose("Archived finalized state bytes", {epoch: finalized.epoch, slot, root: rootHex});
-    } else {
-      // state
-      await this.db.stateArchive.put(finalizedStateOrBytes.slot, finalizedStateOrBytes);
-      // don't delete states before the finalized state, auto-prune will take care of it
-      this.logger.verbose("Archived finalized state", {
-        epoch: finalized.epoch,
-        slot: finalizedStateOrBytes.slot,
-        root: rootHex,
-      });
-    }
+    const state = await this.regen.getCheckpointState(
+      finalized,
+      {dontTransferCache: true},
+      RegenCaller.historicalState
+    );
+    await this.historicalSates?.storeHistoricalState(state.slot, state.serialize());
```

Comment:
be careful with this memory allocation as it could cause `gc` to run for > 1s
need to use `BufferPool` and an util similar to https://github.com/ChainSafe/lodestar/blob/20c18ad85d6cb5e37d4980446d4f608ce9169c24/packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts#L766

in the future I'll implement `serializeInto()` api in ssz to make it more convenient

Reply:
Should we duplicate this logic into utility or expose an interface from the regen?